File: 3ºano/PC/jogomedia.erl
-module(jogomedia).
-export(start)

start(N,Media)-> spawn(fun()-> jogomedia_loop(N,Media,0,0)end).

partida(Pid) ->
    Pid !{participa,self()},
    receive 
        ok -> ok
    end.

aposta(Pid,Numero)->
    Pid !{aposta,Numero,self()},
    receive
        {resultado,R}-> R
    end.

jogomedia_loop(N,Media,Soma,Quantidade)->
    receive
        {participa,From} ->
            From ! ok,
            jogomedia_loop(N,Media,Soma,Quantidade)
        {aposta,Numero,self()} ->
            NovaSoma = Soma + Numero,
            NovaQuantidade = Quantidade +1,
            Res = NovaSoma/NovaQuantidade,
            if
                Res =:= Media ->
                    From !{resultado,true},
                    jogomedia_loop(N,Media,NovaSoma,NovaQuantidade);
                true ->
                    From !{resultado,false}
                    jogomedia_loop(N,Media,NovaSoma,NovaQuantidade);
                end.
end

File: 3rd year/PC/jogomedia.java
interface Jogo{
    Partida participa();
}
interface Partida{
    boolean aposta(int n, int media);
}

class JogoImpl implements Jogo{
    private int numero_jogadores = 0;
    private PartidaImpl partida;
    private boolean timerAtivo = false;
    public synchronized Partida participa(){
        if(numero_jogadores == 0){
            partida = new PartidaImpl();
            iniciar_timer(partida);
        }
        numero_jogadores++;
        return partida;
    }
    public synchronized void iniciar_timer(PartidaImpl partida){
        if(timerAtivo){
            return;
        }
        timerAtivo = true;
        new Thread(()-> {
            try{
                Thread.sleep(60000);
                partida.iniciarPartida();
            }catch(InterruptedException e){}
        }).start();
    }
}
class PartidaImpl implements Partida{
    private int soma = 0;
    private int quantidade = 0;
    private boolean ativa = false;
    public synchronized void iniciarPartida() {
        ativa = true;
    }
    public synchronized boolean aposta(int n, int media){
        if(!ativa){
            return false;
        }
        soma+=n;
        quantidade++;
        int res = soma/quantidade;
        if(res == media){
            return true;
        }
        return false;
    }
}
